src/saveload/ser.rs

```diff
@@ -0,0 +1,169 @@
+use std::fmt::Display;
+use std::marker::PhantomData;
+
+use serde::ser::{self, Serialize, SerializeSeq, Serializer};
+
+use {Entities, FetchMut, Join, ReadStorage, WriteStorage};
+
+use super::{Components, EntityData, Storages};
+use super::marker::{Marker, MarkerAllocator};
+
+/// Serialize components from specified storages via `SerializableComponent::save`
+/// of all marked entities with provided serializer.
+/// When the component gets serialized with `SerializableComponent::save` method
+/// closure passed in `ids` argument mark unmarked `Entity` and they get serialized recursively.
+/// For serializing without such recursion see `serialize` function.
+pub fn serialize_recursive<'a, M, E, T, S>(
+    entities: &Entities<'a>,
+    storages: &<T as Storages<'a>>::ReadStorages,
+    markers: &mut WriteStorage<'a, M>,
+    allocator: &mut FetchMut<'a, M::Allocator>,
+    serializer: S,
+) -> Result<S::Ok, S::Error>
+where
+    M: Marker,
+    E: Display,
+    T: Components<M::Identifier, E>,
+    S: Serializer,
+{
+    let mut serseq = serializer.serialize_seq(None)?;
+    let mut to_serialize = (&**entities, &*markers)
+        .join()
+        .map(|(e, m)| (e, *m))
+        .collect::<Vec<_>>();
+    loop {
+        if to_serialize.is_empty() {
+            break;
+        }
+        let mut add = vec![];
+        {
+            let mut ids = |entity| -> Option<M::Identifier> {
+                match markers.get(entity).cloned() {
+                    Some(marker) => Some(marker.id()),
+                    None => {
+                        let marker = allocator.mark(entity, markers);
```

Comment:
this is a bit convoluted for the fact `allocator.mark` modifies the `markers` parameter
could this be rewritten so that we use a simple `Entry` API here, e.g. `match markers.entry(entity) {...}`?

Reply:
`Entry` API is not there yet. PR #274 
It is possible to call `allocator.allocate(entity, None)` here and insert new marker into storage.
But I'd like it to be this way so the allocator is responsible for the marker attached to entity was allocated for this entity.

Reply:
I can rewrite it as 
```rust
let mut ids = |entity| -> Option<M::Identifier> {
    let (marker, new) = allocator.mark(entity, markers);
    if new { add.push(entity); }
    Some(marker.id())
}
```

modifying `MarkerAllocator::mark` to report if new marker was created or not.

Reply:
`MarkerAllocator::mark` then can be rewritten to use `Entry` API

Reply:
Would you like to rewrite this to use the Entry API in this PR or in a follow-up?